### DISCUSSION TITLE
Copy-Paste inside Safari

### DISCUSSION THREAD

**Discussion by demetr88** [eclipse-rap/org.eclipse.rap#255] [Ideas]
Hi developers
I know that the copy-paste function is available for Safari browser but not I have exception (org.eclipse.swt.SWTError: Cannot set data in clipboard) Could you add this function for RAP?
Best regards

**Comment by ifurnadjiev** [comment]
This is a JavaScript error, which means the programmatic set/get of the clipboard with JavaScript failed. Please note that this feature requires security context (HTTPS) or manually granted permission. Both JavaScript functions, used by RAP are supported in Safari since version 13.1. See:

[1] https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/readText
[2] https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText

**Reply by demetr88** [reply]
My code
try {
            Clipboard clipboard = new Clipboard(Display.getCurrent());
            clipboard.setContents(new Object[]{text}, new Transfer[]{TextTransfer.getInstance()});
            clipboard.dispose();
            infoCopyToClipboard(txtClipboard1, false);
        } catch (SWTError ex) {
            infoCopyToClipboard(txtClipboard3, true);
        }

It has an SWTError-exception for safari.. Can't tell me where I have a mistake
For chrome and firefox it's ok

**Comment by ifurnadjiev** [comment]
The problem is not in your code. Are you running the RAP application over HTTPS? If you run it locally (localhost), did you allow the copy/paste permission in Safari?

**Reply by demetr88** [reply]
yes. https via jetty. 
yes local.
>Did you allow the copy/paste permission in Safari?
hmmm. I didn't allow the copy/paste permission . But I don't think it's necessary, copy-paste works in other services well

**Comment by ifurnadjiev** [comment]
According to this article [1], access to the clipboard permission is not possible in Safari and programmatic clipboard access (without user interaction) is also still not possible.

[1] https://stackoverflow.com/questions/73077159/safari-clipboard-read-permission

**Reply by demetr88** [reply]
Ivan, but I open chatGPT and see button "Copy code" that it works cool for Safari  for copy text to buffer. So it works!
or  am I wrong in my reasoning ? Or in this case, the copy is intended only for the browser not for the server side
But I agree to this option

**Comment by ifurnadjiev** [comment]
By using a BUTTON to interact with the clipboard is probably ok, because it's a "user interaction". In RAP we don't have such a button on the client-side and everything is done with JavaScript WITHOUT "user interaction" driven by the server-side. That's why a specific permission is needed in other browsers. In Safari you don't have access to such permission.

**Reply by demetr88** [reply]
ok I got it. If I bind the button widget with custom JS script will it work?
